### ISSUE TITLE
"error: diffing" refreshing dynamodb.Table.get in Python for table with GSI

### ISSUE THREAD

**Issue by H0R5E** [pulumi/pulumi#16612]
### Describe what happened

Hi, I am having an issue refreshing a stack that has a table retrieved with `dynamodb.Table.get` that is defined in another stack with a GSI. The error I am getting is:

```
aws:dynamodb:Table (bb-4947ngyj-tenant-table):
    error: diffing urn:pulumi:dev::bluebox-tenant-4947ngyj-database-faoz5w2N::aws:dynamodb/table:Table::bb-4947ngyj-tenant-table: all attributes must be indexed. Unused attributes: ["GSIPK"]
```

### Sample program

For a given `tenant_id`, the table is defined in the first stack as follows: 

```python
tenant_table_name = f"bb-{tenant_id}-tenant-table"
tenant_table_gsi_name = tenant_table_name + "-gsi"
tenant_table_attributes = [
    {"name": "GSIPK",
     "type": "S"}
]
tenant_table_gsis = [
    {"name": tenant_table_gsi_name,
     "hash_key": "GSIPK",
     "non_key_attributes": ['databaseStatus',
                            'engine',
                            'lastError',
                            'lastMessage',
                            'lastUpdated',
                            'lastWarning',
                            'latestConfigVersion',
                            'thingMode',
                            'thingName'],
     "projection_type": "INCLUDE"}
]
tenant_table = make_table(tenant_table_name,
                          "PK",
                          attributes=tenant_table_attributes,
                          global_secondary_indexes=tenant_table_gsis)
```

`make_table` is a helper function that returns a `aws.dynamodb.Table` which is referenced in the second stack like this:

```python
tenant_table = aws.dynamodb.Table.get(
        f"bb-{tenant_id}-tenant-table",
        tenant_table_id)
```

where `tenant_table_id` is provided by `tenant_table.name`.

### Log output

I'm not sure how to do this, as I am using the automation API in a container in AWS.

### Affected Resource(s)

dynamodb.Table

### Output of `pulumi about`

I'm not sure how to do this, as I am using the automation API in a container in AWS. I print some info at the start of the run such as:

```
  | 2024-07-09T15:50:39.307Z | Python: 3.9.19 (main, Jun 27 2024, 00:03:35)
  | 2024-07-09T15:50:39.307Z | [GCC 12.2.0]; Pulumi: 3.122.0
  | 2024-07-09T15:50:39.307Z | successfully initialized stack
  | 2024-07-09T15:50:39.307Z | installing plugins...
  | 2024-07-09T15:50:39.307Z | plugins installed
  | 2024-07-09T15:50:39.307Z | setting up config
  | 2024-07-09T15:50:39.307Z | config set
  | 2024-07-09T15:50:39.307Z | refreshing stack...
```

### Additional context

I'm not having any issues refreshing or updating the stack where the table is defined.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by Frassle** [comment]
Does this work on latest pulumi if you set `PULUMI_ENABLE_LEGACY_REFRESH_DIFF=1`?

**Comment by Frassle** [comment]
If the above works then this should also be fixed by https://github.com/pulumi/pulumi/pull/16544 which will be in the next release. Which is targeted to release later today.

**Comment by flostadler** [comment]
Yes, just confirmed it's working with `PULUMI_ENABLE_LEGACY_REFRESH_DIFF=1`. Thanks @Frassle.

@H0R5E if you can set that environment variable you can use this as a workaround, otherwise downgrade to v3.119.0 in the meantime.

**Comment by H0R5E** [comment]
@Frassle, @flostadler yes, I can set that flag. I'll give it a try this afternoon and confirm. Many thanks for your help!

**Comment by H0R5E** [comment]
@flostadler, I can confirm that setting `PULUMI_ENABLE_LEGACY_REFRESH_DIFF=1` fixes the issue for me.

**Comment by justinvp** [comment]
Closing this as a duplicate of #16401. This is now fixed in [v3.123.0](https://github.com/pulumi/pulumi/releases/tag/v3.123.0) (the release is currently rolling out).
